buildSrc/src/main/java/org/elasticsearch/gradle/NoticeTask.java

```diff
@@ -0,0 +1,153 @@
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.gradle;
+
+import org.gradle.api.DefaultTask;
+import org.gradle.api.tasks.InputFile;
+import org.gradle.api.tasks.InputFiles;
+import org.gradle.api.tasks.OutputFile;
+import org.gradle.api.tasks.TaskAction;
+
+import static org.apache.commons.io.FileUtils.readFileToString;
+
+import java.io.IOException;
+import java.io.File;
+import java.nio.file.DirectoryStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.*;
+import java.util.stream.Collectors;
+
+/**
+ * A task to create a notice file which includes dependencies' notices.
+ */
+public class NoticeTask extends DefaultTask {
+    private File inputFile = getProject().getRootProject().file("NOTICE.txt");
+    private File outputFile = new File(getProject().getBuildDir(), "notices/" + getName() + "/NOTICE.txt");
+    /**
+     * Directories to include notices from
+     */
+    private List<File> licensesDirs = new ArrayList<>();
+
+    @InputFile
+    public File getInputFile() {
+        return inputFile;
+    }
+
+    public void setInputFile(File inputFile) {
+        this.inputFile = inputFile;
+    }
+
+    @OutputFile
+    public File getOutputFile() {
+        return outputFile;
+    }
+
+    public void setOutputFile(File outputFile) {
+        this.outputFile = outputFile;
+    }
+
+    public void licensesDir(File licensesDir) {
+        licensesDirs.add(licensesDir);
+    }
+
+    @InputFiles
+    public List<File> getLicensesDirs() {
+        return Collections.unmodifiableList(this.licensesDirs.stream()
+            .map(file -> new File(file.toString())).collect(Collectors.toList()));
+    }
+
+    /**
+     * Add notices from the specified directory.
+     */
+    public NoticeTask() {
+        setDescription("Create a notice file from dependencies");
+        // Default licenses directory is ${projectDir}/licenses (if it exists)
+        File licensesDir = new File(getProject().getProjectDir(), "licenses");
+
+        if (licensesDir.exists()) {
+            licensesDirs.add(licensesDir);
+        }
+    }
+
+    @TaskAction
+    public void generateNotice() throws IOException {
+        final StringBuilder output = new StringBuilder();
+
+        output.append(readFileToString(this.inputFile,"UTF-8"));
+        output.append("\n\n");
+
+        // This is a map rather than a set so that the sort order is the 3rd
+        // party component names, unaffected by the full path to the various files
+        final Map<String, File> seen = this.getFilesToAppend(licensesDirs);
```

Comment:
Replaced the stream that re-threw a checked exception with a function to make it easier to read.